src/modalities/utils/verify_tokenization_constistency.py

```diff
@@ -0,0 +1,198 @@
+            "num_cpus": num_cpus,
+            "eod_token": eod_token,
+            "processing_batch_size": 10,
+            "raw_samples_queue_size": 300,
+            "processed_samples_queue_size": 300,
+        },
+        "tokenizer": {**tokenizer_config},
+    }
+
+    pack_encoded_data(config_dict=tokenization_config_dict)
+
+
+def _verify_index(src_path: Path, index_path: Path):
+    with open(src_path, "rb") as f:
+        jsonl_binary_string = f.read()
+
+    with open(src_path, "rb") as f:
+        binary_string_list = f.readlines()
+
+    with open(src_path, "r", encoding="utf-8") as f:
+        string_list = f.readlines()
+
+    with open(index_path, "rb") as f:
+        jsonl_index = pickle.load(f)
+
+    for i, (offset, length) in tqdm.tqdm(enumerate(jsonl_index), desc="Verifying index"):
+        # check that the index works correctly on the binary data
+        binary_string = binary_string_list[i]
+        if binary_string_list[i].endswith(b"\n"):
+            binary_string = binary_string[:-1]
+        assert jsonl_binary_string[offset : offset + length] == binary_string
+
+        # check that string when encoded with utf-8 matches the binary data
+        string = string_list[i]
+        if string.endswith("\n"):
+            string = string[:-1]
+        assert jsonl_binary_string[offset : offset + length] == string.encode("utf-8")
+
+
+def _verify_pbin(
+    src_path: Path,
+    pbin_path: Path,
+    eod_token_id: int,
+    tokenizer: Callable[[str], list[int]],
+    jsonl_text_key: str,
+):
+    dataset = PackedMemMapDatasetBase(raw_data_path=pbin_path, sample_key="text", load_index=True)
+
+    with open(src_path, "r", encoding="utf-8") as f:
+        string_list = f.readlines()
+    string_list_tokenized = [tokenizer(json.loads(string)[jsonl_text_key]) for string in string_list]
+
+    for i in tqdm.tqdm(range(len(dataset)), desc="Verifying pbin"):
+        pbin_sample = dataset[i]["text"]
+        recomputed_sample = string_list_tokenized[i]
+
+        # make sure that only the last token is the eod token
+        # and that the second last token is not the eod token
+        assert pbin_sample[-1] == eod_token_id
+        assert pbin_sample[-2] != eod_token_id
+
+        # we need to check if tokenizer addas the eod token as
+        # some tokenizers don't add the eod token at the end of the string
+        # whereas modalities always adds the eod token at the end of the string
+        if recomputed_sample[-1] != eod_token_id:
+            if i == 0:
+                warnings.warn("The tokenizer does not add the eod token at the end of the string!")
+            assert len(pbin_sample) - 1 == len(recomputed_sample)
+            assert all(pbin_sample[:-1] == recomputed_sample)
+        else:
+            assert len(pbin_sample) == len(recomputed_sample)
+            assert all(pbin_sample == recomputed_sample)
+
+
+def build_hf_tokenization_components(tokenizer_path_or_name: str, eod_token: str):
```

Comment:
The two functions `build_hf_tokenization_components` and `build_sp_tokenization_components` seem to only be used in `tests/dataloader/test_end_to_end_indexation_and_tokenization.py`. Wouldn't it make more sense to move them there?

Reply:
This is true and I though about it. The reason why I kept them there is because we can easily test any SP and HF tokenizer e.g., on a custom dataset that we want to verify works correctly.